File: backend/src/models/product.ts
import mongoose, { Schema, Document, Model } from 'mongoose';

interface IProduct extends Document {
  name: string;
  description: string;
  price: number;
  user_id: mongoose.Types.ObjectId;
}

const productSchema: Schema<IProduct> = new mongoose.Schema<IProduct>({
  name: { type: String, required: true, unique: true, collation: { locale: 'en', strength: 2 } },
  description: { type: String, required: true },
  price: { type: Number, required: true },
  user_id: { type: Schema.Types.ObjectId, required: true, ref: 'User' },
});

const Product: Model<IProduct> = mongoose.model<IProduct>('Product', productSchema);

export { Product, IProduct };

File: frontend/lib/services/product.dart
// product class
class Product {
  final String productId;
  final String name;
  final double price;
  final String description;
  final String userId;

  Product({required this.productId, required this.name, required this.price, required this.description, required this.userId});

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      productId: json['_id'],
      name: json['name'],
      price: json['price'].toDouble(),
      description: json['description'],
      userId: json['user_id'],
    );
  }
}
